### PR TITLE
3604 - 3

### PR DESCRIPTION
**Pull request by MargaretOm**
## Номер
3604

## Номер задания
3

## Ссылка на видео с демо
https://drive.google.com/open?id=1hB_pHX6_5OoFGXpjOmWGBSWIpeYk6ixE

### CODE REVIEW

File: 3604/3/config/environment.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require 'bundler'
+Bundler.require
+require 'rack-flash'
+
+ActiveRecord::Base.establish_connection(
+  adapter: 'sqlite3',
+  database: 'foo.sqlite3'
```
Comment (ermaac) [inline 3604/3/config/environment.rb]: not sure that it is good name for database
you can set it based on environment (development, staging, production). Check RACK_ENV variable for this. You can leave development as a default

File: 3604/3/application/models/review.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class Review < ActiveRecord::Base
```
Comment (ermaac) [inline 3604/3/application/models/review.rb]: this model misses validations

File: 3604/3/application/models/place.rb [hunk 3]
```diff
@@ -0,0 +1,3 @@
+class Place < ActiveRecord::Base
```
Comment (ermaac) [inline 3604/3/application/models/place.rb]: this model misses validations

File: 3604/3/application/controllers/user_controller.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      session_start!
+      session[:user_id] = @user.id
+      redirect '/main'
+    else
+      flash[:error] = I18n.t(:incorrect_sign_in)
+      redirect '/sign_in'
+    end
+  end
+
+  get '/sign_out' do
```
Comment (ermaac) [inline 3604/3/application/controllers/user_controller.rb]: this is semantic issue. you are destroying session, not fetching/reading. Choose DELETE or POST http method for this (preferable is delete)

**Comment by ermaac** [comment]
Please add action for routes which you do not handle (e.g /cats, /dogs etc.)
You can either redirect to places page or display separate page which tells that page is not found

File: 3604/3/application/controllers/place_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    @place = Place.find_by(params[:id])
+    @reviews = Review.where(place_id: params[:id])
+    @stars = stars(@reviews) if @reviews.count.positive?
+    erb :place
+  end
+
+  post '/review/:id' do
+    @review = Review.new(grade: params[:grade], text: params[:text],
+                         place_id: params[:id], user_id: session[:user_id])
+    @review.save
```
Comment (ermaac) [inline 3604/3/application/controllers/place_controller.rb]: What will be if review is not valid?

File: 3604/3/application/controllers/place_controller.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    redirect '/place/' + params[:id]
```
Comment (ermaac) [inline 3604/3/application/controllers/place_controller.rb]: please use string interpolation

File: 3604/3/application/public/css/main.css [hunk 7]
```diff
@@ -0,0 +1,137 @@
+body {
+  background: rgb(206, 192, 192);
```
[the remark this answers is not in the document]
Comment (MargaretOm) [reply 3604/3/application/public/css/main.css]: What about colors (white, black and etc.)? Use hex format?

File: 3604/3/application/controllers/place_controller.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  private
+
+  def stars(reviews)
```
[the remark this answers is not in the document]
Comment (MargaretOm) [reply 3604/3/application/controllers/place_controller.rb]: @ermaac but if i want use this method stars in future with other hashes

File: 3604/3/application/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+class User < ActiveRecord::Base
+  has_many :reviews
+  validates_presence_of :name, :email, :password_hash
+  validates_uniqueness_of :email
+  validate :email_can_be_valid
+
+  def email_can_be_valid
```
Comment (ermaac) [inline 3604/3/application/models/user.rb]: please replace can with must

**Pull request merged**